bitbots_vision/bitbots_vision/yoeo_vision.py

```diff
@@ -0,0 +1,257 @@
+#! /usr/bin/env python3
+from typing import Dict, Union, List
+
+import os
+import rclpy
+from rclpy.node import Node
+from ament_index_python.packages import get_package_share_directory
+from rcl_interfaces.msg import SetParametersResult
+from copy import deepcopy
+from cv_bridge import CvBridge
+from threading import Lock
+from sensor_msgs.msg import Image
+from bitbots_vision.vision_modules import yoeo_handler, ros_utils
+from bitbots_vision.vision_modules.yoeo_vision_components import IVisionComponent, CameraCapCheckComponent, \
+    YOEOBallDetectionComponent, YOEOObstacleDetectionComponent, YOEOGoalpostDetectionComponent, \
+    YOEOLineDetectionComponent, YOEOFieldBoundaryDetectionComponent, YOEOFieldDetectionComponent, DebugImageComponent
+
+from .yoeo_params import gen
+
+logger = rclpy.logging.get_logger('bitbots_vision')
+
+try:
+    from profilehooks import profile, \
+        timecall  # Profilehooks profiles certain functions in you add the @profile or @timecall decorator.
+except ImportError:
+    profile = lambda x: x
+    logger.info("No Profiling available")
+
+
+class YOEOVision(Node):
+    """
+    The Vision is the main ROS-node for handling all tasks related to image processing.
+
+    This class defines the whole image processing pipeline, which uses the modules from the `vision_modules`.
+    It also handles the dynamic reconfiguration of the bitbots_vision.
+    """
+
+    def __init__(self):
+        super().__init__('bitbots_vision')
+
+        self._package_path = get_package_share_directory('bitbots_vision')
+
+        logger.info(f"Entering {self.__class__.__name__} constructor")
+
+        self._cv_bridge = CvBridge()
+
+        self._config: Dict = {}
+
+        self._sub_image = None
+
+        # Reconfigure data transfer variable
+        self._transfer_reconfigure_data = None
+        self._transfer_reconfigure_data_mutex = Lock()
+
+        # Image transfer variable
+        self._transfer_image_msg = None
+        self._transfer_image_msg_mutex = Lock()
```

Comment:
This is probably/hopefully not necessary anymore, as in ROS2 the vision apparently now runs in a single threaded event-loop.

Reply:
@Flova what's your opinion on this?

Reply:
I see this similar to Jan. To be sure a single threaded executer could be used explicitly. This is different to ROS1 where the default executor was multi threaded.